File: baekjoon/05-BruteForce/python/3085.py
# https://www.acmicpc.net/problem/3085 사탕 게임 (python3: 4608ms / pypy3: 500ms)
import sys
sys.stdin = open('input.txt', 'r')
input = sys.stdin.readline
'''
N의 크기가 50이므로 브루트포스 탐색이 가능하다
  1.가로로 양옆의 사탕을 바꾸고 연속된 최대 개수를 카운트한다
  2.세로로 상하의 사탕을 바꾸고 연속된 최대 개수를 카운트한다
'''
n = int(input())
candy = [list(input().rstrip()) for _ in range(n)]
ans = 1

# 몇 개 먹을 수 있는지 찾는 함수
def count_candy():
    global ans    
    # 행에서의 최대 개수 카운트
    for i in range(n):
        cnt = 1
        for j in range(n-1):
            if candy[i][j] == candy[i][j+1]:
                cnt += 1
                ans = max(ans, cnt)
            else:
                cnt = 1
    # 열에서의 최대 개수 카운트
    for i in range(n):
        cnt = 1
        for j in range(n-1):
            if candy[j][i] == candy[j+1][i]:
                cnt += 1
                ans = max(ans, cnt)
            else:
                cnt = 1

# [모든 인접한 두 자리 뒤집어보고 최대 개수 찾기]
# 가로 뒤집기
for i in range(n):
    for j in range(n-1):
        candy[i][j], candy[i][j+1] = candy[i][j+1], candy[i][j] # 양옆 위치를 바꾸고
        count_candy()
        candy[i][j], candy[i][j+1] = candy[i][j+1], candy[i][j] # 위치 원상 복귀
# 세로 뒤집기
for i in range(n):
    for j in range(n-1):
        candy[j][i], candy[j+1][i] = candy[j+1][i], candy[j][i]
        count_candy()
        candy[j][i], candy[j+1][i] = candy[j+1][i], candy[j][i]

print(ans)

File: baekjoon/05-BruteForce/js/3085.js
// 백준 제출 시, readFileSync('/dev/stdin') 으로 변경할 것
// https://www.acmicpc.net/problem/3085 사탕 게임 (440ms)
const fs = require("fs");
const input = fs.readFileSync("input.txt").toString().trim().split("\n");

const n = Number(input[0]);
const candy = new Array(n).fill([]);
for (let i = 0; i < n; i++) {
    tmp = [];
    for (let j = 0; j < n; j++) {
        tmp.push(input[i + 1][j]);
    }
    candy[i] = tmp;
}
let ans = 1;

// 사탕을 먹을 수 있는 최대 개수 카운트 함수
function countCandy() {
    // 행
    for (let i = 0; i < n; i++) {
        let cnt = 1;
        for (let j = 0; j < n - 1; j++) {
            if (candy[i][j] == candy[i][j + 1]) {
                cnt++;
                ans = Math.max(ans, cnt);
            } else {
                cnt = 1;
            }
        }
    }
    // 열
    for (let i = 0; i < n; i++) {
        let cnt = 1;
        for (let j = 0; j < n - 1; j++) {
            if (candy[j][i] == candy[j + 1][i]) {
                cnt++;
                ans = Math.max(ans, cnt);
            } else {
                cnt = 1;
            }
        }
    }
}

function swapCandy(i, j, flag) {
    if (flag == true) {
        // 가로 바꾸기 (열의 원소 바꾸기)
        const tmp = candy[i][j];
        candy[i][j] = candy[i][j + 1];
        candy[i][j + 1] = tmp;
    } else {
        // 세로 바꾸기 (행의 원소 바꾸기)
        const tmp = candy[j][i];
        candy[j][i] = candy[j + 1][i];
        candy[j + 1][i] = tmp;
    }
}

// [모든 인접한 두 자리 바꾸면서 최대 개수 찾기]
// 가로 바꾸기 (열의 원소 바꾸기)
for (let i = 0; i < n; i++) {
    for (let j = 0; j < n - 1; j++) {
        swapCandy(i, j, true); // 바꾸기
        countCandy();
        swapCandy(i, j, true); // 원위치
    }
}
// 세로 바꾸기 (행의 원소 바꾸기)
for (let i = 0; i < n; i++) {
    for (let j = 0; j < n - 1; j++) {
        swapCandy(i, j, false); // 바꾸기
        countCandy();
        swapCandy(i, j, false); // 원위치
    }
}

console.log(ans);
